pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/scheduler/UnloadScheduler.java

```diff
@@ -122,33 +141,40 @@ public synchronized void execute() {
                     log.info("Only 1 broker available: no load shedding will be performed. Skipping.");
                     return CompletableFuture.completedFuture(null);
                 }
-                final UnloadDecision unloadDecision = namespaceUnloadStrategy
+                final Set<UnloadDecision> decisions = namespaceUnloadStrategy
                         .findBundlesForUnloading(context, recentlyUnloadedBundles, recentlyUnloadedBrokers);
                 if (debugMode) {
                     log.info("[{}] Unload decision result: {}",
-                            namespaceUnloadStrategy.getClass().getSimpleName(), unloadDecision.toString());
+                            namespaceUnloadStrategy.getClass().getSimpleName(), decisions);
                 }
-                if (unloadDecision.getUnloads().isEmpty()) {
+                if (decisions.isEmpty()) {
                     if (debugMode) {
                         log.info("[{}] Unload decision unloads is empty. Skipping.",
                                 namespaceUnloadStrategy.getClass().getSimpleName());
                     }
                     return CompletableFuture.completedFuture(null);
                 }
                 List<CompletableFuture<Void>> futures = new ArrayList<>();
-                unloadDecision.getUnloads().forEach((broker, unload) -> {
+                long asyncOpTimeoutMs = conf.getNamespaceBundleUnloadingTimeoutMs();
+                decisions.forEach(decision -> {
+                    Unload unload = decision.getUnload();
                     log.info("[{}] Unloading bundle: {}", namespaceUnloadStrategy.getClass().getSimpleName(), unload);
                     futures.add(unloadManager.waitAsync(channel.publishUnloadEventAsync(unload), unload.serviceUnit(),
-                                    conf.getNamespaceBundleUnloadingTimeoutMs(), TimeUnit.MILLISECONDS)
+                                    decision, asyncOpTimeoutMs, TimeUnit.MILLISECONDS)
                             .thenAccept(__ -> {
-                                recentlyUnloadedBundles.put(unload.serviceUnit(), System.currentTimeMillis());
-                                recentlyUnloadedBrokers.put(unload.sourceBroker(), System.currentTimeMillis());
+                        recentlyUnloadedBundles.put(unload.serviceUnit(), System.currentTimeMillis());
+                        recentlyUnloadedBrokers.put(unload.sourceBroker(), System.currentTimeMillis());
                     }));
                 });
-                return FutureUtil.waitForAll(futures).exceptionally(ex -> {
-                    log.error("[{}] Namespace unload has exception.",
-                            namespaceUnloadStrategy.getClass().getSimpleName(), ex);
-                    return null;
+                return FutureUtil.waitForAll(futures).whenComplete((__, ex) -> {
```

Comment:
Do we have any reason not to sync-wait here? I think we better wait for the unloads to complete and return.

Reply:
updated.